From 460938a7c50a985bc50e9c073f74038f65c9f7a3 Mon Sep 17 00:00:00 2001
From: Hariharan Basuthkar <quic_hbasuthk@quicinc.com>
Date: Tue, 4 Jun 2024 00:50:40 +0530
Subject: [PATCH] wifi: ath12k: Validate power mode in nl80211_parse_chandef

In nl80211_parse_chandef, validate the 6 GHz power mode received
from the NL attribute.

Signed-off-by: Hariharan Basuthkar <quic_hbasuthk@quicinc.com>
---
 net/wireless/nl80211.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -3624,6 +3624,16 @@ int nl80211_parse_chandef(struct cfg8021
 		if (attrs[NL80211_ATTR_CENTER_FREQ2])
 			chandef->center_freq2 =
 				nla_get_u32(attrs[NL80211_ATTR_CENTER_FREQ2]);
+
+		if (control_freq >= MHZ_TO_KHZ(5945) && control_freq <= MHZ_TO_KHZ(7125)) {
+			err = cfg80211_validate_freq_width_for_pwr_mode(&rdev->wiphy,
+									chandef,
+									mode);
+			if (err) {
+				NL_SET_ERR_MSG(extack, "Invalid frequency for power mode");
+				return err;
+			}
+		}
 	}
 
 	if (info->attrs[NL80211_ATTR_RU_PUNCT_BITMAP]) {
